src/common/internet.rs

```diff
@@ -0,0 +1,310 @@
+/// * IPv4-to-IPv6 address
+/// * IPv6 address
+/// * Tor address (only 3rd version is supported)
+///
+/// NB: we are using `TorPublicKeyV3` instead of `OnionAddressV3`, since
+/// `OnionAddressV3` keeps cehcksum and other information wich can be
+/// reconstructed from `TorPublicKeyV3`. The 2-byte checksum in `OnionAddressV3`
+/// is designed for human-readable part that checks that the address was typed
+/// in correctly. In computer-stored digital data it may be deterministically
+/// regenerated and does not add any additional security.
+///
+/// Tor addresses are distinguished by the fact that last 16 bits
+/// must be set to 0
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum InetAddr {
+    IPv4(Ipv4Addr),
+    IPv6(Ipv6Addr),
+    #[cfg(feature="use-tor")]
+    Tor(TorPublicKeyV3),
+}
+
+impl InetAddr {
+    pub fn get_ip6(&self) -> Option<Ipv6Addr> {
+        match self {
+            InetAddr::IPv4(ipv4_addr) => Some(ipv4_addr.to_ipv6_mapped()),
+            InetAddr::IPv6(ipv6_addr) => Some(*ipv6_addr),
+            #[cfg(feature="use-tor")]
+            _ => None,
+        }
+    }
+
+    pub fn from_32bit_encoding(data: [u8; TORV3_PUBLIC_KEY_LENGTH]) -> Option<Self> {
+        match data {
+            d if d[0..28] == [0u8; 28] => {
+                let mut a = [0u8; 4];
+                a.clone_from_slice(&d[28..]);
+                Some(InetAddr::IPv4(Ipv4Addr::from(a)))
+            },
+            d if d[0..24] == [0u8; 16] => {
+                let mut a = [0u8; 16];
+                a.clone_from_slice(&d[16..]);
+                Some(InetAddr::IPv6(Ipv6Addr::from(a)))
+            },
+            #[cfg(feature="use-tor")]
+            d  => TorPublicKeyV3::from_bytes(&d).map(|pk| InetAddr::Tor(pk)).ok(),
+        }
+    }
+
+    pub fn get_32bit_encoding(&self) -> [u8; TORV3_PUBLIC_KEY_LENGTH] {
+        let mut buf = [0u8; 32];
+        match self {
+            InetAddr::IPv4(ipv4_addr) => buf[24..].copy_from_slice(&ipv4_addr.octets()),
+            InetAddr::IPv6(ipv6_addr) => buf[16..].copy_from_slice(&ipv6_addr.octets()),
+            #[cfg(feature="use-tor")]
+            InetAddr::Tor(tor_addr) => buf = tor_addr.to_bytes(),
+        }
+        buf
+    }
+}
+
+impl fmt::Display for InetAddr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            // TODO:
+            InetAddr::IPv4(addr) => write!(f, "{}", addr),
+            InetAddr::IPv6(addr) => write!(f, "{}", addr),
+            #[cfg(feature="use-tor")]
+            InetAddr::Tor(addr) => write!(f, "{}", addr),
+        }
+    }
+}
+
+#[cfg(feature="use-tor")]
+impl TryFrom<InetAddr> for IpAddr {
+    type Error = ();
+    #[inline]
+    fn try_from(addr: InetAddr) -> Result<Self, Self::Error> {
+        Ok(match addr {
+            InetAddr::IPv4(addr) => IpAddr::V4(addr),
+            InetAddr::IPv6(addr) => IpAddr::V6(addr),
+            InetAddr::Tor(addr) => Err(())?,
+        })
+    }
+}
```

Comment:
I would keep `TryFrom` always implemented, even if the `use-tor` feature is disabled: Ideally if i was a user of this library i would prefer to write code that doesn't break if i disable a feature, unless i was explicitly trying to use that feature.

Reply:
Hm, good point... And if one does not use Tor, they do not need this structures, BTW